theano/tensor/sort.py

```diff
@@ -206,3 +219,311 @@ def argsort(a, axis=-1, kind='quicksort', order=None):
+                fn_argmax(x, axis=axis), axis)
+            idx2 = tuple(
+                np.arange(s).reshape(
+                    (s,) + (1,) * (ndim - i - 1)
+                    ) if i != axis else zi for i, s in enumerate(x.shape))
+            zv = x[idx2]
+            return zv, zi.astype(idx_dtype)
+        else:
+            zi = np.expand_dims(
+                fn_argmax(x, axis=axis), axis)
+            return zi.astype(idx_dtype)
+
+    if x.shape[axis] == abs(k):
+        if not op.return_indices:
+            return x.copy()
+        else:
+            l = axis
+            r = ndim - l
+            reps = list(x.shape)
+            reps[axis] = 1
+            zi = np.arange(abs(k), dtype=idx_dtype)
+            zi = zi.reshape((1,) * l + (k,) + (1,) * (r - 1))
+            zi = np.tile(zi, reps)
+            if op.return_values:
+                return x.copy(), zi
+            else:
+                return zi
+
+    idx = [slice(None)] * ndim
+    idx[axis] = (slice(-k, None) if k > 0 else slice(-k))
+
+    if not op.return_indices:
+        zv = np.partition(x, -k, axis=axis)[idx]
+        return zv
+    elif op.return_values:
+        zi = np.argpartition(x, -k, axis=axis)[idx]
+        idx2 = tuple(
+            np.arange(s).reshape(
+                (s,) + (1,) * (ndim - i - 1)
+                ) if i != axis else zi for i, s in enumerate(x.shape))
+        zv = x[idx2]
+        return zv, zi.astype(idx_dtype)
+    else:
+        zi = np.argpartition(x, -k, axis=axis)[idx]
+        return zi.astype(idx_dtype)
+
+
+class TopKOp(theano.Op):
+    """
+    Operations related to finding k-largest elements.
+
+    Parameters
+    ----------
+    axis: integer
+        Defaults to ``-1``.
+        The axis to perform the operation. Must be in range ``[-ndim, ndim)``, where
+        ``ndim`` is the dimensionality of input tensor.
+
+    idx_dtype: string
+        Specify output dtype for indices, defaults to ``int64``, must be integer type.
+
+
+    Notes
+    -----
+    - By default, this Op give two outputs: values and indices. However optimizer may
+      remove a certain output if not needed.
+
+    - Computing gradient is only possible when both values and indices are computed in
+      forward pass.
+
+    - If the top-k-th value is not unique, we cannot guarantee the output indices being
+      deterministically chosen.
+
+    See Also
+    --------
+    topk
+    argtopk
+    argtopk_and_topk
+
+    """
+
+    # TODO more params
+    '''
+    sorted: bool
+        Defaults to ``False``
+
+        If True, the result array would be incremental-sorted.
+
+    only_top_kth: bool
+        Defaults to ``False``
+
+        If ``True``, will only find one exact top k-th element on given axis.
+
+    '''
+
+    # TODO c_code
+    # TODO add opt, if k==1, use max/min reduce
+    #      also if k is axis size, just copy input tensor
+    # TODO add opt to merge argtopk / topk, or split topk_and_argtopk when only
```

Comment:
I did a few tests. For a local opt, replacement must be same type as old output, and can't be `None`.

What's the right fix for this? Should I just plug a orphan variable?